In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
import pandas as pd
import numpy as np
import matplotlib
import json

In [3]:
from data_engineering import de

In [4]:
df = pd.read_csv("data/movie_data.csv")

/Users/daiglechris/Git/movie_ds_model/venv/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3063: DtypeWarning: Columns (0,1,2,3,4,5,6,7,8,9,10,13,14,15,16,19) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [5]:
# Initial analysis of dataframe
null_0 = pd.DataFrame(df.isnull().sum()).reset_index()
null_0.rename(columns={'index': 'factor', 0: 'num_null_0'}, inplace=True)
null_0.sort_values(by='num_null_0', axis=0, ascending=False, inplace=True)
df_info = {'obs_0':df.shape[0],
           'fac_0':df.shape[1],
           'null_0': null_0}
# Transformations to more accurately reflect data integrity issues
numeric_cols = ['budget', 'popularity', 'revenue',
                'runtime', 'vote_average', 'vote_count']
for col in numeric_cols:
    df[col] = df[col].apply(lambda x: pd.to_numeric(x, downcast='float', errors='coerce'))
df['id'] = df['id'].apply(lambda x: pd.to_numeric(x, errors='coerce'))
df['release_date'] = df['release_date'].apply(lambda x: pd.to_datetime(x, errors='coerce'))
# Remove observations related to response variables
df = df.dropna(how='any', subset=['vote_count', 'vote_average', 'revenue'])
# Dataframe of the number of missing values after null removals
trans_0_null = pd.DataFrame(df.isnull().sum()).reset_index()
trans_0_null.rename(columns={'index': 'factor', 0: 'num_null_1'}, inplace=True)
null_1 = null_0.join(trans_0_null.set_index('factor'), on='factor')
# Update dataframe information dictionary
df_info['obs_1'] = df.shape[0]
df_info['fac_1'] = df.shape[1]
df_info['null_1'] = null_1

In [6]:
print("Initial observations: {:,}".format(df_info['obs_0']))
print("Number of observations after removing missing 'vote_count', 'vote_average', 'revenue': {:,}".format(df_info['obs_1']))

Initial observations: 133,744
Number of observations after removing missing 'vote_count', 'vote_average', 'revenue': 28,477


In [7]:
for col in df.columns:
    print(col)
    print(df[col].dtypes)
    print("{:.2%} null".format(df[col].isna().sum() / df.shape[0]))
    display(df[col].head())
    print("+" * len(col))

budget
float64
0.00% null


0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: budget, dtype: float64

++++++
genres
object
0.00% null


0                       [{"id": 35, "name": "Comedy"}]
1                    [{"id": 12, "name": "Adventure"}]
2    [{"id": 10749, "name": "Romance"}, {"id": 18, ...
3                       [{"id": 35, "name": "Comedy"}]
4                                                   []
Name: genres, dtype: object

++++++
homepage
object
85.90% null


0                        NaN
1                        NaN
2                        NaN
3                        NaN
4    http://www.nwdfilms.com
Name: homepage, dtype: object

++++++++
id
float64
0.00% null


0     3924.0
1     6124.0
2     8773.0
3    21624.0
4    25449.0
Name: id, dtype: float64

++
original_language
object
0.00% null


0    en
1    de
2    fr
3    pa
4    en
Name: original_language, dtype: object

+++++++++++++++++
original_title
object
0.00% null


0                               Blondie
1                   Der Mann ohne Namen
2                   L'amour à vingt ans
3                      Jatts in Golmaal
4    New World Disorder 9: Never Enough
Name: original_title, dtype: object

++++++++++++++
overview
object
1.61% null


0    Blondie and Dagwood are about to celebrate the...
1    Der Mann ohne Namen is a German adventure movi...
2    Love at Twenty unites five directors from five...
3    Jatts in Golmal is an Comedy based movie. In w...
4    Gee Atherton ripping the Worlds course the day...
Name: overview, dtype: object

++++++++
popularity
float64
0.00% null


0    3.067
1    0.845
2    3.701
3    0.600
4    1.212
Name: popularity, dtype: float64

++++++++++
production_companies
object
0.00% null


0    [{"id": 5, "logo_path": "/71BqEFAF4V3qjjMPCpLu...
1                                                   []
2    [{"id": 38936, "logo_path": null, "name": "Uly...
3                                                   []
4                                                   []
Name: production_companies, dtype: object

++++++++++++++++++++
production_countries
object
0.00% null


0    [{"iso_3166_1": "US", "name": "United States o...
1            [{"iso_3166_1": "DE", "name": "Germany"}]
2    [{"iso_3166_1": "DE", "name": "Germany"}, {"is...
3                                                   []
4                                                   []
Name: production_countries, dtype: object

++++++++++++++++++++
release_date
datetime64[ns]
0.39% null


0   1938-11-30
1   1921-01-01
2   1962-06-22
3   2003-02-21
4   2008-12-08
Name: release_date, dtype: datetime64[ns]

++++++++++++
revenue
float64
0.00% null


0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: revenue, dtype: float64

+++++++
runtime
float64
0.60% null


0     70.0
1    420.0
2    110.0
3      NaN
4     69.0
Name: runtime, dtype: float64

+++++++
spoken_languages
object
0.00% null


0             [{"iso_639_1": "en", "name": "English"}]
1                                                   []
2    [{"iso_639_1": "de", "name": "Deutsch"}, {"iso...
3                                                   []
4             [{"iso_639_1": "en", "name": "English"}]
Name: spoken_languages, dtype: object

++++++++++++++++
status
object
0.00% null


0    Released
1    Released
2    Released
3    Released
4    Released
Name: status, dtype: object

++++++
tagline
object
54.43% null


0                                     NaN
1                                     NaN
2    The Intimate Secrets of Young Lovers
3                                     NaN
4                                     NaN
Name: tagline, dtype: object

+++++++
title
object
0.00% null


0                               Blondie
1         Peter Voss, Thief of Millions
2                        Love at Twenty
3                      Jatts in Golmaal
4    New World Disorder 9: Never Enough
Name: title, dtype: object

+++++
vote_average
float64
0.00% null


0    6.6
1    0.0
2    6.9
3    0.0
4    4.5
Name: vote_average, dtype: float64

++++++++++++
vote_count
float64
0.00% null


0     4.0
1     0.0
2    27.0
3     0.0
4     2.0
Name: vote_count, dtype: float64

++++++++++
keywords
object
0.00% null


0                  [{"id": 190801, "name": "blondie"}]
1                                                   []
2                                                   []
3                                                   []
4    [{"id": 6075, "name": "sport"}, {"id": 10192, ...
Name: keywords, dtype: object

++++++++


## Data Engineering Strategy
For each column in the dataframe, here's some info and my plan to handle them:
1. **budget**: float64, 0% null
> *ALREADY CONVERTED TO FLOAT AND COERCED NAN*
2. **genres**: object, 0% null, array holding dictionary, ex: `x = [{"id": 35, "name": "Comedy"}]`
> *DRAW VALUE FROM "name" FROM EACH ELEMENT FROM ARRAY, ASSIGN NAN IF EMPTY*
3. **homepage**: object, 85.9% null
> *DROP: TOO MUCH MISSING*
4. **id**: float64, 0% null
> *DROP: NO INFORMATION*
5. **original_language**: object, 0% null
> *ENFORCE STRING*
6. **original_title**: object, 0% null
> *ENFORCE STRING*
7. **overview**: object, 1.61% null
> *ENFORCE STRING*
8. **popularity**: float64, 0% null
9. **production_companies**: object, 0% null, array holding dictionary
> *DRAW VALUE FROM "name" FROM EACH ELEMENT FROM ARRAY, ASSIGN NAN IF EMPTY*
10. **production_countries**: object, 0% null, array holding dictionaries
> *DRAW VALUE FROM "iso_3166_1" FROM EACH ELEMENT FROM ARRAY, ASSIGN NAN IF EMPTY*
11. **release_date**: datetime64, 0.39% null
> *ALREADY CONVERTED TO DATETIME AND FILLED NAN*
12. **revenue**: float64, 0% null
> *RESPONSE VARIABLE*
13. **runtime**: float64, 0.6% null
> *ALREADY CONVERTED TO FLOAT AND COERCED NAN*
14. **spoken_languages**: object, 0% null, array of dictionaries
> *Drop and see if valuable after using only original_language*<br>
> If brought back in...<br>
> *DRAW VALUE FROM "iso_639_1" FROM EACH ELEMENT FROM ARRAY, ASSIGN NAN IF EMPTY, CAPITALIZE*
15. **status**: object, 0% null
> *CHECK IF BINARY INDICATOR*
16. **tagline**: object, 54.43% null
> *FIND RELATIONSHIP BETWEEN ORIGINAL_LANGUAGE, PRODUCTION_COUNTRIES, AND STATUS*
17. **title**: object, 0% null
18. **vote_average**: float64, 0% null
> *PREVIOUSLY CONVERTED TO FLOAT AND DROPPED MISSING*
19. **vote_count**: float64, 0% null
> *PREVIOUSLY CONVERTED TO FLOAT AND DROPPED MISSING*
20. **keywords**: object, 0% null, array holding dictionaries
> *DRAW VALUE FROM "name" FROM EACH ELEMENT FROM ARRAY, ASSIGN NAN IF EMPTY*

In [8]:
df['genres'] = de.string_array_clean(df_col=df['genres'], element='name')
df.drop(['homepage', 'id'], axis=1, inplace=True)
df['original_language'] = df['original_language'].astype(str)
df['original_title'] = df['original_title'].astype(str)
df['overview'] = df['overview'].astype(str)
df['production_companies'] = de.string_array_clean(df_col=df['production_companies'], element='name')
df['production_countries'] = de.string_array_clean(df_col=df['production_countries'], element='iso_3166_1')
df['spoken_languages'] = de.string_array_clean(df_col=df['spoken_languages'], element='iso_639_1')
df['tagline'] = df['tagline'].astype(str)
df['title'] = df['title'].astype(str)
df['keywords'] = de.string_array_clean(df_col=df['keywords'], element='name')

In [9]:
df.head()

budget            genres original_language  \
0     0.0          [Comedy]                en   
1     0.0       [Adventure]                de   
2     0.0  [Romance, Drama]                fr   
3     0.0          [Comedy]                pa   
4     0.0               NaN                en   

                       original_title  \
0                             Blondie   
1                 Der Mann ohne Namen   
2                 L'amour à vingt ans   
3                    Jatts in Golmaal   
4  New World Disorder 9: Never Enough   

                                            overview  popularity  \
0  Blondie and Dagwood are about to celebrate the...       3.067   
1  Der Mann ohne Namen is a German adventure movi...       0.845   
2  Love at Twenty unites five directors from five...       3.701   
3  Jatts in Golmal is an Comedy based movie. In w...       0.600   
4  Gee Atherton ripping the Worlds course the day...       1.212   

                                production_companies  production_countries  \
0                                [Columbia Pictures]                  [US]   
1                                                NaN                  [DE]   
2  [Ulysse Productions, Unitec Films, Cinesecolo,...  [DE, FR, IT, JP, PL]   
3                                                NaN                   NaN   
4                                                NaN                   NaN   

  release_date  revenue  runtime      spoken_languages    status  \
0   1938-11-30      0.0     70.0                  [en]  Released   
1   1921-01-01      0.0    420.0                   NaN  Released   
2   1962-06-22      0.0    110.0  [de, fr, it, ja, pl]  Released   
3   2003-02-21      0.0      NaN                   NaN  Released   
4   2008-12-08      0.0     69.0                  [en]  Released   

                                tagline                               title  \
0                                   nan                             Blondie   
1                                   nan       Peter Voss, Thief of Millions   
2  The Intimate Secrets of Young Lovers                      Love at Twenty   
3                                   nan                    Jatts in Golmaal   
4                                   nan  New World Disorder 9: Never Enough   

   vote_average  vote_count                keywords  
0           6.6         4.0               [blondie]  
1           0.0         0.0                     NaN  
2           6.9        27.0                     NaN  
3           0.0         0.0                     NaN  
4           4.5         2.0  [sport, mountain bike]

In [24]:
df[df['original_title'].apply(lambda x: len(x)) == 0]

budget                                             genres  \
29             0.0  [{"id": 16, "name": "Animation"}, {"id": 878, ...   
41      41000000.0  [{"id": 18, "name": "Drama"}, {"id": 10402, "n...   
44      15000000.0  [{"id": 35, "name": "Comedy"}, {"id": 878, "na...   
81      10000000.0  [{"id": 53, "name": "Thriller"}, {"id": 80, "n...   
107       310607.0  [{"id": 18, "name": "Drama"}, {"id": 27, "name...   
...            ...                                                ...   
133497         0.0           [{"id": 878, "name": "Science Fiction"}]   
133584         0.0  [{"id": 28, "name": "Action"}, {"id": 37, "nam...   
133613         0.0  [{"id": 18, "name": "Drama"}, {"id": 10749, "n...   
133652         0.0  [{"id": 18, "name": "Drama"}, {"id": 53, "name...   
133686  18000000.0  [{"id": 80, "name": "Crime"}, {"id": 18, "name...   

       homepage       id original_language original_title  \
29          NaN     30.0                ja         彼女の想いで   
41          NaN     65.0                en         8 Mile   
44          NaN     68.0                en         Brazil   
81          NaN    107.0                en         Snatch   
107         NaN    136.0                en         Freaks   
...         ...      ...               ...            ...   
133497      NaN  43230.0                en         Kronos   
133584      NaN  43329.0                en         Apache   
133613      NaN  43358.0                en         Carrie   
133652      NaN  43397.0                en         Caught   
133686      NaN  43434.0                en         Carlos   

                                                 overview  popularity  \
29      Koji Morimato’s animated science fiction short...       5.554   
41      The setting is Detroit in 1995. The city is di...      23.692   
44      Low-level bureaucrat Sam Lowry escapes the mon...      12.990   
81      There are two overlapping stories. One is the ...      24.026   
107     A circus' beautiful trapeze artist agrees to m...      10.358   
...                                                   ...         ...   
133497  Scientists investigate a huge meteor that cras...       2.718   
133584  Following the surrender of Geronimo, Massai, t...       6.867   
133613  Carrie's dreams of adventure in the big city a...       2.142   
133652  It was Leonora Eames' childhood dream come tru...       2.968   
133686  The story of Venezuelan revolutionary, Ilich R...       6.344   

                                     production_companies  \
29      [{"id": 11671, "logo_path": "/iRdYQdn1eunwV3qA...   
41      [{"id": 24, "logo_path": null, "name": "Mikona...   
44      [{"id": 10214, "logo_path": null, "name": "Emb...   
81      [{"id": 3287, "logo_path": "/bz6GbCQQXGNE56LTW...   
107     [{"id": 21, "logo_path": "/mjofSXiHpG5t6KYmU4l...   
...                                                   ...   
133497  [{"id": 13973, "logo_path": null, "name": "Reg...   
133584  [{"id": 30724, "logo_path": null, "name": "Hec...   
133613  [{"id": 4, "logo_path": "/fycMZt242LVjagMByZOL...   
133652  [{"id": 21, "logo_path": "/mjofSXiHpG5t6KYmU4l...   
133686  [{"id": 11394, "logo_path": null, "name": "Fil...   

                                     production_countries  ...      revenue  \
29                [{"iso_3166_1": "JP", "name": "Japan"}]  ...          0.0   
41      [{"iso_3166_1": "US", "name": "United States o...  ...  242875078.0   
44      [{"iso_3166_1": "GB", "name": "United Kingdom"...  ...          0.0   
81      [{"iso_3166_1": "GB", "name": "United Kingdom"...  ...   83557872.0   
107     [{"iso_3166_1": "US", "name": "United States o...  ...          0.0   
...                                                   ...  ...          ...   
133497  [{"iso_3166_1": "US", "name": "United States o...  ...          0.0   
133584  [{"iso_3166_1": "US", "name": "United States o...  ...          0.0   
133613  [{"iso_3166_1": "US", "name": "United States o...  ...          0.0   
13